File: src/core/cloudflare-manager.ts
/**
 * Read-safe Cloudflare Pages REST client — the agent-reachable half of the
 * integration, and the mirror of `vercel-manager.ts`.
 *
 * Like {@link ./git-manager.js GitManager}, nothing here changes what is
 * deployed: retry / rollback and the environment-variable writes live in
 * `cloudflare-actions.ts` behind the dashboard's guarded route.
 *
 * Two shapes of Cloudflare's API drive most of what looks odd below:
 *
 * 1. **Everything is account-scoped.** Vercel's `teamId` is optional; a
 *    Cloudflare account id is in the path of every Pages call, so an unscoped
 *    client can answer `viewer()` and `listAccounts()` and nothing else.
 * 2. **Projects are addressed by name, deployments only within a project.**
 *    There is no global deployment endpoint, which is why every deployment read
 *    here takes the project name.
 */

import type { ProviderFetch } from "./providers/egress.js";

/** Cloudflare wraps every response; `result` is the payload. */
interface CloudflareEnvelope<T> {
  success?: boolean;
  errors?: { code?: number; message?: string }[];
  result?: T;
}

export interface CloudflareUser {
  id: string;
  email?: string | null;
  username?: string | null;
}

export interface CloudflareAccount {
  id: string;
  name: string;
}

/** A project's Git connection, when it was imported from a repository. */
export interface CloudflareSource {
  type: string;
  owner?: string;
  repoName?: string;
  productionBranch?: string;
}

export interface CloudflareProject {
  /** Cloudflare's own UUID. Not what the API addresses the project by — see {@link name}. */
  uuid: string;
  /** The path segment every project-scoped call uses. */
  name: string;
  framework?: string | null;
  productionBranch?: string | null;
  createdAt?: number;
  updatedAt?: number;
  source?: CloudflareSource;
  /**
   * Build settings. Null means "not set", which Cloudflare renders as its own
   * default — distinct from the field being absent, though in practice the list
   * and single-project endpoints return the same shape, unlike Vercel's.
   */
  buildCommand?: string | null;
  destinationDir?: string | null;
  rootDir?: string | null;
  /** Per-environment variables, embedded in the project rather than listed separately. */
  envVars: CloudflareEnvVar[];
}

/**
 * One variable, already merged across the environments it appears in.
 *
 * Cloudflare stores these as a `{ NAME: { type, value } }` map per environment,
 * so the same key in production and preview is two entries there and one here.
 * `value` is present only for `plain_text`; a `secret_text` value never reads
 * back over the API at all.
 */
export interface CloudflareEnvVar {
  key: string;
  environments: string[];
  type: "plain_text" | "secret_text";
  value?: string;
}

export interface CloudflareDeploymentStage {
  name: string;
  status: string;
}

export interface CloudflareDeployment {
  id: string;
  shortId?: string;
  projectName: string;
  url: string | null;
  environment: string;
  stage: CloudflareDeploymentStage;
  isSkipped: boolean;
  createdAt: number;
  modifiedAt?: number;
  aliases: string[];
  branch?: string;
  commitHash?: string;
  commitMessage?: string;
}

export interface CloudflareBuildLogLine {
  id: string;
  createdAt: number;
  text: string;
}

export interface CloudflareDomain {
  name: string;
  status: string;
  createdAt?: number;
  /** Set when validation is outstanding — the DNS record the user must add. */
  validation?: { method?: string; status?: string; txtName?: string; txtValue?: string };
  errorMessage?: string;
}

export class CloudflareApiError extends Error {
  constructor(
    message: string,
    public readonly status: number,
    public readonly path: string,
    /** Cloudflare's own numeric error code, when it sent one. */
    public readonly code?: number,
  ) {
    super(message);
    this.name = "CloudflareApiError";
  }
}

/**
 * Cloudflare keys an imported project by owner and repository name rather than
 * by a URL, so the "repo url" this provider matches on is `owner/repo`.
 *
 * The contract calls this a repo *URL* because that is what Vercel wants; what
 * it actually means is "the string this vendor recognizes a repository by", and
 * only {@link CloudflareManager.findProjectByRepo} ever interprets it.
 */
export function cloudflareRepoUrl(remoteUrl: string): string | null {
  const trimmed = remoteUrl.trim().replace(/\.git$/, "").replace(/\/$/, "");
  const ssh = /^[\w.-]+@[^:]+:(?<path>.+)$/.exec(trimmed);
  const path = ssh?.groups?.path ?? trimmed.replace(/^https?:\/\/[^/]+\//, "");
  const parts = path.split("/").filter(Boolean);
  if (parts.length < 2) return null;
  return parts.slice(-2).join("/").toLowerCase();
}

export class CloudflareManager {
  constructor(
    private readonly token: string,
    private readonly accountId?: string,
    private readonly baseUrl = "https://api.cloudflare.com/client/v4",
    /** The provider's scoped `fetch`; global `fetch` when built outside the registry. */
    private readonly fetchImpl?: ProviderFetch,
  ) {}

  /**
   * The signed-in account.
   *
   * `/user` needs a token carrying User Details Read, which a project-scoped
   * API token typically does not have. Rather than fail the whole dashboard
   * over an identity nicety, a rejected `/user` falls back to the token's own
   * verification and the selected account's name.
   */
  async viewer(): Promise<CloudflareUser> {
    try {
      const user = await this.request<CloudflareUser & { email?: string }>("/user");
      return { id: user.id, email: user.email ?? null, username: user.username ?? user.email };
    } catch (error) {
      if (!(error instanceof CloudflareApiError) || error.status >= 500) throw error;
      return this.tokenViewer(error);
    }
  }

  /**
   * Identity for a token that cannot read `/user`. `/user/tokens/verify` is
   * answerable by every API token, so a failure there is a genuine auth error
   * and is left to propagate.
   */
  private async tokenViewer(cause: CloudflareApiError): Promise<CloudflareUser> {
    const verified = await this.request<{ id?: string; status?: string }>("/user/tokens/verify");
    if (verified.status && verified.status !== "active") {
      throw new CloudflareApiError(
        `This Cloudflare API token is ${verified.status}.`,
        401,
        "/user/tokens/verify",
        cause.code,
      );
    }
    const accounts = await this.listAccounts().catch(() => []);
    const account = accounts.find((entry) => entry.id === this.accountId) ?? accounts[0];
    return {
      id: verified.id ?? this.accountId ?? "",
      email: null,
      username: account?.name ?? "Cloudflare API token",
    };
  }

  async listAccounts(): Promise<CloudflareAccount[]> {
    const accounts = await this.request<{ id: string; name?: string }[]>("/accounts?per_page=50");
    return accounts.map((account) => ({ id: account.id, name: account.name ?? account.id }));
  }

  /**
   * Every project in the account.
   *
   * Cloudflare has no search or repository filter on this endpoint, so both are
   * applied here. That is a page of extra bytes, not a page of extra requests —
   * the list endpoint returns whole project objects either way.
   */
  async listProjects(
    opts: { search?: string; repoUrl?: string; limit?: number } = {},
  ): Promise<CloudflareProject[]> {
    const perPage = Math.min(Math.max(opts.limit ?? 50, 1), 100);
    const raw = await this.request<RawProject[]>(
      `/accounts/${this.account()}/pages/projects?per_page=${perPage}`,
    );
    let projects = raw.map(normalizeProject);
    if (opts.search) {
      const needle = opts.search.toLowerCase();
      projects = projects.filter((project) => project.name.toLowerCase().includes(needle));
    }
    if (opts.repoUrl) {
      const wanted = opts.repoUrl.toLowerCase();
      projects = projects.filter((project) => repoSlug(project) === wanted);
    }
    return projects;
  }

  /** The project a repository deploys, matched on `owner/repo`. */
  async findProjectByRepo(repoUrl: string): Promise<CloudflareProject | undefined> {
    return (await this.listProjects({ repoUrl, limit: 100 }))[0];
  }

  async getProject(name: string): Promise<CloudflareProject> {
    return normalizeProject(
      await this.request<RawProject>(
        `/accounts/${this.account()}/pages/projects/${encodeURIComponent(name)}`,
      ),
    );
  }

  async listDeployments(opts: {
    projectName: string;
    env?: "production" | "preview";
    limit?: number;
  }): Promise<CloudflareDeployment[]> {
    const params = new URLSearchParams({ per_page: String(Math.min(opts.limit ?? 20, 100)) });
    if (opts.env) params.set("env", opts.env);
    const raw = await this.request<RawDeployment[]>(
      `${this.projectPath(opts.projectName)}/deployments?${params}`,
    );
    return raw.map(normalizeDeployment);
  }

  async getDeployment(projectName: string, deploymentId: string): Promise<CloudflareDeployment> {
    return normalizeDeployment(
      await this.request<RawDeployment>(
        `${this.projectPath(projectName)}/deployments/${encodeURIComponent(deploymentId)}`,
      ),
    );
  }

  /**
   * The deployment that currently serves production, which Cloudflare records
   * on the project rather than on the deployment. A rollback moves it, so it
   * cannot be inferred from "the newest successful production build".
   */
  async canonicalDeploymentId(projectName: string): Promise<string | undefined> {
    const raw = await this.request<RawProject>(
      `/accounts/${this.account()}/pages/projects/${encodeURIComponent(projectName)}`,
    );
    return raw.canonical_deployment?.id ?? raw.latest_deployment?.id;
  }

  async deploymentBuildLogs(
    projectName: string,
    deploymentId: string,
    limit = 500,
  ): Promise<CloudflareBuildLogLine[]> {
    const data = await this.request<{ data?: { ts?: string; line?: string }[] }>(
      `${this.projectPath(projectName)}/deployments/${encodeURIComponent(deploymentId)}/history/logs`,
    );
    const lines = (data.data ?? [])
      .map((entry, index) => ({
        id: `${entry.ts ?? index}-${index}`,
        createdAt: epochMs(entry.ts) ?? 0,
        text: (entry.line ?? "").trimEnd(),
      }))
      .filter((line) => line.text.length > 0);
    // Cloudflare returns the whole build; the caller's limit is the tail of it,
    // which is where a failure's cause is.
    return lines.slice(-limit);
  }

  /** Variables for the project, keys and environments only — values are dropped. */
  async listEnv(projectName: string): Promise<CloudflareEnvVar[]> {
    const project = await this.getProject(projectName);
    return project.envVars.map(({ value: _value, ...rest }) => rest);
  }

  /**
   * One variable's value.
   *
   * Cloudflare returns `plain_text` values in the project payload but never a
   * `secret_text` one — the value is write-only. So this can answer for the
   * former and must say so plainly for the latter, rather than returning an
   * empty string the UI would render as "this secret is blank".
   */
  async getEnvValue(projectName: string, key: string): Promise<string> {
    const project = await this.getProject(projectName);
    const variable = project.envVars.find((entry) => entry.key === key);
    if (!variable) throw new Error(`No variable named "${key}" on this project.`);
    if (variable.type === "secret_text") {
      throw new Error("Cloudflare never returns a secret's value. Set a new one to change it.");
    }
    return variable.value ?? "";
  }

  async listDomains(projectName: string): Promise<CloudflareDomain[]> {
    const raw = await this.request<RawDomain[]>(`${this.projectPath(projectName)}/domains`);
    return raw.map(normalizeDomain);
  }

  /** The dashboard page for a deployment, which Cloudflare does not return. */
  inspectorUrl(projectName: string, deploymentId: string): string | undefined {
    if (!this.accountId) return undefined;
    return `https://dash.cloudflare.com/${this.accountId}/pages/view/${encodeURIComponent(projectName)}/${encodeURIComponent(deploymentId)}`;
  }

  private projectPath(projectName: string): string {
    return `/accounts/${this.account()}/pages/projects/${encodeURIComponent(projectName)}`;
  }

  /**
   * The account every Pages path needs. Throwing here rather than sending a
   * request with `undefined` in the URL is what turns "no account chosen" into
   * a sentence the scope switcher answers.
   */
  private account(): string {
    if (!this.accountId) {
      throw new Error("Choose a Cloudflare account before reading its Pages projects.");
    }
    return this.accountId;
  }

  private request<T>(path: string, opts?: { method?: string; body?: unknown }): Promise<T> {
    return cloudflareRequest<T>(
      { token: this.token, baseUrl: this.baseUrl, fetch: this.fetchImpl },
      path,
      opts,
    );
  }
}

export interface CloudflareRequestAuth {
  token: string;
  baseUrl?: string;
  /**
   * The provider's scoped `fetch` (see `providers/egress.ts`). Absent means
   * global `fetch` — the case for a manager built directly in a test. Production
   * clients come from `cloudflare-context.ts`, which always supplies one.
   */
  fetch?: ProviderFetch;
}

/**
 * The one place a Cloudflare API call is made. Shared with
 * `cloudflare-actions.ts` so auth and error shaping stay identical across the
 * read and write halves — the read/write boundary is which module exposes an
 * operation, not which one can reach the network.
 */
export async function cloudflareRequest<T>(
  auth: CloudflareRequestAuth,
  path: string,
  opts?: { method?: string; body?: unknown },
): Promise<T> {
  const url = path.startsWith("http")
    ? path
    : `${auth.baseUrl ?? "https://api.cloudflare.com/client/v4"}${path}`;
  const headers: Record<string, string> = {
    Authorization: `Bearer ${auth.token}`,
    Accept: "application/json",
  };
  const init: RequestInit = { method: opts?.method ?? "GET", headers };
  if (opts?.body !== undefined) {
    headers["Content-Type"] = "application/json";
    init.body = JSON.stringify(opts.body);
  }

  const response = await (auth.fetch ?? fetch)(url, init);
  const body = (await response.json().catch(() => ({}))) as CloudflareEnvelope<T>;
  // Cloudflare can answer 200 with `success: false`, so the envelope decides,
  // not the status line.
  if (!response.ok || body.success === false) {
    const failure = body.errors?.[0];
    throw new CloudflareApiError(
      failure?.message ?? `Cloudflare returned ${response.status} for ${path}`,
      response.status,
      path,
      failure?.code,
    );
  }
  return body.result as T;
}

interface RawEnvVarEntry {
  type?: string;
  value?: string;
}

interface RawDeploymentConfig {
  env_vars?: Record<string, RawEnvVarEntry | null> | null;
}

interface RawProject {
  id: string;
  name: string;
  framework?: string | null;
  production_branch?: string | null;
  created_on?: string;
  subdomain?: string;
  build_config?: {
    build_command?: string | null;
    destination_dir?: string | null;
    root_dir?: string | null;
  } | null;
  deployment_configs?: Record<string, RawDeploymentConfig | null> | null;
  source?: {
    type?: string;
    config?: { owner?: string; repo_name?: string; production_branch?: string };
  } | null;
  canonical_deployment?: { id?: string } | null;
  latest_deployment?: { id?: string; modified_on?: string } | null;
}

interface RawDeployment {
  id: string;
  short_id?: string;
  project_name?: string;
  url?: string | null;
  environment?: string;
  is_skipped?: boolean;
  created_on?: string;
  modified_on?: string;
  aliases?: string[] | null;
  latest_stage?: { name?: string; status?: string } | null;
  deployment_trigger?: {
    metadata?: { branch?: string; commit_hash?: string; commit_message?: string };
  } | null;
}

interface RawDomain {
  name: string;
  status?: string;
  created_on?: string;
  validation_data?: {
    method?: string;
    status?: string;
    txt_name?: string;
    txt_value?: string;
    error_message?: string;
  } | null;
  verification_data?: { status?: string; error_message?: string } | null;
}

/** Cloudflare timestamps are RFC 3339 strings; everything downstream wants epoch ms. */
function epochMs(value: string | undefined | null): number | undefined {
  if (!value) return undefined;
  const parsed = Date.parse(value);
  return Number.isNaN(parsed) ? undefined : parsed;
}

const repoSlug = (project: CloudflareProject): string | undefined =>
  project.source?.owner && project.source.repoName
    ? `${project.source.owner}/${project.source.repoName}`.toLowerCase()
    : undefined;

/**
 * Flattens the per-environment `{ NAME: { type, value } }` maps into one entry
 * per key, carrying every environment it appears in. A key that is plain text
 * in one environment and a secret in another is reported as a secret, because
 * the stricter of the two is the one that governs whether a value may be read.
 */
function normalizeEnvVars(
  configs: Record<string, RawDeploymentConfig | null> | null | undefined,
): CloudflareEnvVar[] {
  const merged = new Map<string, CloudflareEnvVar>();
  for (const [environment, config] of Object.entries(configs ?? {})) {
    for (const [key, entry] of Object.entries(config?.env_vars ?? {})) {
      if (!entry) continue;
      const type = entry.type === "plain_text" ? "plain_text" : "secret_text";
      const existing = merged.get(key);
      if (existing) {
        existing.environments.push(environment);
        if (type === "secret_text") existing.type = "secret_text";
        if (existing.value === undefined) existing.value = entry.value;
        continue;
      }
      merged.set(key, { key, environments: [environment], type, value: entry.value });
    }
  }
  return [...merged.values()].sort((a, b) => a.key.localeCompare(b.key));
}

function normalizeProject(project: RawProject): CloudflareProject {
  const build = project.build_config ?? {};
  return {
    uuid: project.id,
    name: project.name,
    framework: project.framework ?? null,
    productionBranch: project.production_branch ?? null,
    createdAt: epochMs(project.created_on),
    updatedAt: epochMs(project.latest_deployment?.modified_on) ?? epochMs(project.created_on),
    source: project.source?.type
      ? {
          type: project.source.type,
          owner: project.source.config?.owner,
          repoName: project.source.config?.repo_name,
          productionBranch:
            project.source.config?.production_branch ?? project.production_branch ?? undefined,
        }
      : undefined,
    // Cloudflare always sends the whole `build_config`, so a null here really
    // does mean "not set" rather than "not read".
    buildCommand: build.build_command ?? null,
    destinationDir: build.destination_dir ?? null,
    rootDir: build.root_dir ?? null,
    envVars: normalizeEnvVars(project.deployment_configs),
  };
}

function normalizeDeployment(deployment: RawDeployment): CloudflareDeployment {
  const trigger = deployment.deployment_trigger?.metadata ?? {};
  return {
    id: deployment.id,
    shortId: deployment.short_id,
    projectName: deployment.project_name ?? "",
    // Cloudflare returns a full URL where the contract wants a bare hostname.
    url: deployment.url ? deployment.url.replace(/^https?:\/\//, "").replace(/\/$/, "") : null,
    environment: deployment.environment ?? "preview",
    stage: {
      name: deployment.latest_stage?.name ?? "queued",
      status: deployment.latest_stage?.status ?? "idle",
    },
    isSkipped: deployment.is_skipped ?? false,
    createdAt: epochMs(deployment.created_on) ?? 0,
    modifiedAt: epochMs(deployment.modified_on),
    aliases: (deployment.aliases ?? []).map((alias) =>
      alias.replace(/^https?:\/\//, "").replace(/\/$/, ""),
    ),
    branch: trigger.branch,
    commitHash: trigger.commit_hash,
    commitMessage: trigger.commit_message,
  };
}

function normalizeDomain(domain: RawDomain): CloudflareDomain {
  const validation = domain.validation_data ?? undefined;
  return {
    name: domain.name,
    status: domain.status ?? "pending",
    createdAt: epochMs(domain.created_on),
    validation: validation
      ? {
          method: validation.method,
          status: validation.status,
          txtName: validation.txt_name,
          txtValue: validation.txt_value,
        }
      : undefined,
    errorMessage: validation?.error_message ?? domain.verification_data?.error_message,
  };
}

File: crates/nomoreide-core/src/cloudflare_manager.rs
//! Read-safe Cloudflare Pages REST client — the Rust counterpart of
//! `src/core/cloudflare-manager.ts`.
//!
//! Like `vercel_manager`, this module deliberately contains no operation that
//! changes what is deployed: retry and rollback live on the write side.
//!
//! Two things shape everything here. Cloudflare wraps every answer in a
//! `{ success, errors, result }` envelope, so a failure carries a message even
//! when the status does not; and Pages paginates everything, ten at a time,
//! with no way to ask for more — so a listing is a walk rather than a request.

use serde_json::Value;

use crate::providers::api_base::{provider_api_base, provider_api_host};
use crate::providers::egress::ProviderEgress;

const API_BASE: &str = "https://api.cloudflare.com/client/v4";
/// Pages' own page size. Not configurable at the vendor.
const PAGE_SIZE: usize = 10;
/// How far a listing walks before giving up. Ten pages is a hundred projects,
/// which is more than a dashboard can show and far more than a person has.
const MAX_PAGES: usize = 10;
/// Accounts are asked for in one request; a token with more than fifty is not
/// a case the account picker could help with anyway.
const ACCOUNT_PAGE_SIZE: usize = 50;

/// Cloudflare's API, or the loopback stand-in an environment override names.
pub fn api_base() -> String {
    provider_api_base("NOMOREIDE_CLOUDFLARE_API_BASE", API_BASE)
}

fn egress() -> ProviderEgress {
    ProviderEgress::new("cloudflare", vec![provider_api_host(&api_base())])
}

#[derive(Debug, Clone)]
pub struct CloudflareApiError {
    pub message: String,
    pub status: u16,
}

impl std::fmt::Display for CloudflareApiError {
    fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(formatter, "{}", self.message)
    }
}

/// The refusal every account-scoped read starts with.
///
/// Pages addresses everything as `/accounts/<id>/…`, and `/accounts` answers
/// with an empty list rather than an error for a token that lacks
/// `Account Settings: Read` — so a working Pages token can arrive here with
/// nothing to pick from, and the message has to cover both.
pub const NO_ACCOUNT: &str = "Choose a Cloudflare account before reading its Pages projects. If the account list is empty, this API token has no account access — recreate it with Account Resources including your account.";

/// The one place a Cloudflare API call is made.
pub async fn request(token: &str, path: &str) -> Result<Value, CloudflareApiError> {
    let url = if path.starts_with("http") {
        path.to_string()
    } else {
        format!("{}{path}", api_base())
    };

    let client = reqwest::Client::builder()
        .redirect(reqwest::redirect::Policy::none())
        .build()
        .map_err(|error| CloudflareApiError {
            message: format!("Cloudflare request failed: {error}"),
            status: 0,
        })?;
    let bearer = token.to_string();
    let response = egress()
        .send(
            &client,
            &url,
            |client, verb, target| {
                client
                    .request(verb, target)
                    .header("Authorization", format!("Bearer {bearer}"))
                    .header("Accept", "application/json")
            },
            reqwest::Method::GET,
        )
        .await
        .map_err(|error| CloudflareApiError {
            message: format!("Cloudflare request failed: {error}"),
            status: 0,
        })?;

    let status = response.status();
    let text = response.text().await.unwrap_or_default();
    let body: Value = serde_json::from_str(&text).unwrap_or(Value::Null);
    if status.is_success() {
        return Ok(body);
    }
    Err(CloudflareApiError {
        message: api_error_message(&body, status.as_u16(), path),
        status: status.as_u16(),
    })
}

/// Cloudflare puts the useful sentence in `errors[0].message`. When it sends
/// none, the bare status says nothing about what was being asked, so the
/// request is named instead.
fn api_error_message(body: &Value, status: u16, path: &str) -> String {
    body.get("errors")
        .and_then(Value::as_array)
        .and_then(|errors| errors.first())
        .and_then(|error| error.get("message"))
        .and_then(Value::as_str)
        .map(str::to_string)
        .unwrap_or_else(|| format!("Cloudflare returned {status} for {path}"))
}

/// A field the vendor actually sent, in the sense `??` means it: an absent key
/// and an explicit `null` are both "no value", and everything else survives as
/// the JSON it was rather than being narrowed to a string.
fn present(field: Option<&Value>) -> Option<Value> {
    field.filter(|value| !value.is_null()).cloned()
}

/// The identity record, with the keys the vendor had nothing for left out —
/// `JSON.stringify` drops an `undefined` field, and the shape is compared.
fn user_value(id: Option<Value>, email: Option<Value>, username: Option<Value>) -> Value {
    let mut user = serde_json::Map::new();
    for (key, value) in [("id", id), ("email", email), ("username", username)] {
        if let Some(value) = value {
            user.insert(key.into(), value);
        }
    }
    Value::Object(user)
}

/// The `result` of an envelope, as a list.
fn results(body: &Value) -> Vec<Value> {
    body.get("result")
        .and_then(Value::as_array)
        .cloned()
        .unwrap_or_default()
}

pub struct CloudflareManager {
    token: String,
    account_id: Option<String>,
}

impl CloudflareManager {
    pub fn new(token: String, account_id: Option<String>) -> Self {
        Self { token, account_id }
    }

    pub fn account_id(&self) -> Option<&str> {
        self.account_id.as_deref()
    }

    fn require_account(&self) -> Result<&str, CloudflareApiError> {
        self.account_id.as_deref().ok_or(CloudflareApiError {
            message: NO_ACCOUNT.to_string(),
            status: 0,
        })
    }

    /// Who this credential belongs to.
    ///
    /// `/user` answers for a Wrangler OAuth login but is forbidden to a scoped
    /// API token, which is the common case — so a 4xx here is not a failure,
    /// it is the signal to identify the token instead. A 5xx (or a transport
    /// failure, which arrives as status 0) is a real outage and propagates.
    ///
    /// Cloudflare's `username` is an opaque 32-char hex id rather than a
    /// display name, so the email is the only human-readable identity `/user`
    /// carries. Preferring `username` here put a hex blob in the account menu.
    pub async fn viewer(&self) -> Result<Value, CloudflareApiError> {
        let failure = match request(&self.token, "/user").await {
            Ok(body) => {
                let user = body.get("result").cloned().unwrap_or(Value::Null);
                let email = present(user.get("email"));
                let username = email.clone().or_else(|| present(user.get("username")));
                // `email` is reported even when there is none — the reference
                // writes `email ?? null`, so the key is always on the record.
                return Ok(user_value(
                    present(user.get("id")),
                    Some(email.unwrap_or(Value::Null)),
                    username,
                ));
            }
            Err(failure) => failure,
        };
        if failure.status == 0 || failure.status >= 500 {
            return Err(failure);
        }
        self.token_viewer().await
    }

    /// Identity for a token that cannot read `/user`. `/user/tokens/verify` is
    /// answerable by every API token, so a failure *there* is a genuine auth
    /// error and is left to propagate.
    async fn token_viewer(&self) -> Result<Value, CloudflareApiError> {
        let verified = request(&self.token, "/user/tokens/verify")
            .await?
            .get("result")
            .cloned()
            .unwrap_or(Value::Null);
        // Only a token Cloudflare calls something other than `active` is
        // rejected; a response that names no status at all is accepted, the
        // way a missing field always is here.
        if let Some(status) = verified.get("status").and_then(Value::as_str) {
            if status != "active" {
                return Err(CloudflareApiError {
                    message: format!("This Cloudflare API token is {status}."),
                    status: 401,
                });
            }
        }
        let accounts = self.list_accounts().await.unwrap_or_default();
        let account = accounts
            .iter()
            .find(|entry| entry.get("id").and_then(Value::as_str) == self.account_id.as_deref())
            .or_else(|| accounts.first());
        // `listAccounts` reports `name ?? id`, so the fallback to the id is the
        // account's own and not this function's.
        let name = account
            .and_then(|entry| present(entry.get("name")).or_else(|| present(entry.get("id"))));
        Ok(user_value(
            present(verified.get("id"))
                .or_else(|| self.account_id.clone().map(Value::String))
                .or(Some(Value::String(String::new()))),
            Some(Value::Null),
            name.or_else(|| Some(Value::String("Cloudflare API token".into()))),
        ))
    }

    /// Every account this token can see, which is what the scope picker offers.
    pub async fn list_accounts(&self) -> Result<Vec<Value>, CloudflareApiError> {
        let path = format!("/accounts?per_page={ACCOUNT_PAGE_SIZE}");
        Ok(results(&request(&self.token, &path).await?))
    }

    /// Every Pages project in the account, walked a page at a time.
    pub async fn list_projects_raw(&self) -> Result<Vec<Value>, CloudflareApiError> {
        let account = self.require_account()?.to_string();
        let mut projects = Vec::new();
        for page in 1..=MAX_PAGES {
            let path =
                format!("/accounts/{account}/pages/projects?per_page={PAGE_SIZE}&page={page}");
            let batch = results(&request(&self.token, &path).await?);
            let short = batch.len() < PAGE_SIZE;
            projects.extend(batch);
            if short {
                break;
            }
        }
        projects.truncate(MAX_PAGES * PAGE_SIZE);
        Ok(projects)
    }

    pub async fn get_project_raw(&self, name: &str) -> Result<Value, CloudflareApiError> {
        let account = self.require_account()?.to_string();
        let path = format!("/accounts/{account}/pages/projects/{}", segment(name));
        Ok(request(&self.token, &path)
            .await?
            .get("result")
            .cloned()
            .unwrap_or(Value::Null))
    }

    /// A project's deployments, newest first, walked until `limit` are in hand.
    ///
    /// The environment filter is the vendor's own (`env=`), unlike Vercel's
    /// preview case — Pages labels every deployment with one.
    pub async fn list_deployments_raw(
        &self,
        project: &str,
        target: Option<&str>,
        limit: usize,
    ) -> Result<Vec<Value>, CloudflareApiError> {
        let account = self.require_account()?.to_string();
        let filter = target
            .map(|target| format!("?env={}", segment(target)))
            .unwrap_or_default();
        let separator = if filter.is_empty() { '?' } else { '&' };
        let mut deployments: Vec<Value> = Vec::new();
        for page in 1..=MAX_PAGES {
            let path = format!(
                "/accounts/{account}/pages/projects/{}/deployments{filter}{separator}per_page={PAGE_SIZE}&page={page}",
                segment(project)
            );
            let batch = results(&request(&self.token, &path).await?);
            let short = batch.len() < PAGE_SIZE;
            deployments.extend(batch);
            if short || deployments.len() >= limit {
                break;
            }
        }
        deployments.truncate(limit);
        Ok(deployments)
    }

    pub async fn get_deployment_raw(
        &self,
        project: &str,
        deployment: &str,
    ) -> Result<Value, CloudflareApiError> {
        let account = self.require_account()?.to_string();
        let path = format!(
            "/accounts/{account}/pages/projects/{}/deployments/{}",
            segment(project),
            segment(deployment)
        );
        Ok(request(&self.token, &path)
            .await?
            .get("result")
            .cloned()
            .unwrap_or(Value::Null))
    }

    /// Build history for a deployment. Pages serves it as one document; the
    /// live tail is a websocket, which is why there is no `follow` here.
    pub async fn build_logs_raw(
        &self,
        project: &str,
        deployment: &str,
    ) -> Result<Vec<Value>, CloudflareApiError> {
        let account = self.require_account()?.to_string();
        let path = format!(
            "/accounts/{account}/pages/projects/{}/deployments/{}/history/logs",
            segment(project),
            segment(deployment)
        );
        Ok(request(&self.token, &path)
            .await?
            .get("result")
            .and_then(|result| result.get("data"))
            .and_then(Value::as_array)
            .cloned()
            .unwrap_or_default())
    }
}

/// One path segment, encoded the way `encodeURIComponent` encodes it.
///
/// The account id deliberately does not go through this: the reference
/// interpolates it raw, and an id it encoded differently would address a
/// different account.
fn segment(value: &str) -> String {
    urlencoding::encode(value).into_owned()
}

/// The URL Cloudflare keys an imported project by: `owner/repo`, lowercased.
///
/// Pages stores the two halves separately rather than as a URL, so this is what
/// a git remote has to be reduced to before the two can be compared.
pub fn repo_url(remote_url: &str) -> Option<String> {
    let trimmed = remote_url.trim().trim_end_matches('/');
    let trimmed = trimmed.strip_suffix(".git").unwrap_or(trimmed);
    let parts: Vec<&str> = trimmed
        .split(['/', ':'])
        .filter(|part| !part.is_empty())
        .collect();
    if parts.len() < 2 {
        return None;
    }
    Some(parts[parts.len() - 2..].join("/").to_lowercase())
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn an_error_body_without_a_message_names_the_request() {
        assert_eq!(
            api_error_message(&json!({"errors": []}), 404, "/accounts"),
            "Cloudflare returned 404 for /accounts"
        );
        assert_eq!(
            api_error_message(
                &json!({"errors": [{"code": 10000, "message": "Authentication error"}]}),
                403,
                "/accounts"
            ),
            "Authentication error"
        );
    }

    /// Pages keys a project by owner and repo, so a remote in any of git's
    /// spellings has to reduce to the same pair.
    #[test]
    fn a_remote_reduces_to_owner_and_repo() {
        for remote in [
            "https://github.com/Acme/App.git",
            "git@github.com:Acme/App.git",
            "https://github.com/Acme/App/",
            "ssh://git@github.com/Acme/App",
        ] {
            assert_eq!(repo_url(remote).as_deref(), Some("acme/app"), "{remote}");
        }
        assert_eq!(repo_url("app"), None);
    }
}
